In [917]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib import cm
from IPython.display import display,HTML
from datetime import datetime
from dateutil.parser import parse
# reading data files
Ind_last_4years_home = pd.read_csv("C:\\Users\\sushanth\\Desktop\\Ind_last_4years_home.csv", sep=",", header = 0, engine='python', encoding ="utf-8", skipinitialspace=True)
Ind_Aus_host_India_ten_years = pd.read_csv("C:\\Users\\sushanth\\Desktop\\Ind_Aus_host_India_ten_years.csv", sep=",", header = 0, engine='python', encoding ="utf-8", skipinitialspace=True)
Ind_Aus_host_India_selected_venues = pd.read_csv("C:\\Users\\sushanth\\Desktop\\Ind_Aus_host_India_selected_venues.csv", sep=",", header = 0, engine='python', encoding ="utf-8", skipinitialspace=True)
Ind_last_2years = pd.read_csv("C:\\Users\\sushanth\\Desktop\\Ind_last_2years.csv", sep=",", header = 0, engine='python', encoding ="utf-8", skipinitialspace=True)
testCases = pd.read_csv("C:\\Users\\sushanth\\Desktop\\testCases.csv", sep=",", header = 0, engine='python', encoding ="utf-8", skipinitialspace=True)

In [918]:
Ind_last_4years_home.info()
Ind_last_4years_home.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29 entries, 0 to 28
Data columns (total 10 columns):
Unnamed: 0    0 non-null float64
BR            11 non-null float64
Bat           29 non-null object
Ground        29 non-null object
Margin        29 non-null object
Opposition    29 non-null object
Result        29 non-null object
Team          29 non-null object
Toss          29 non-null object
Year          29 non-null int64
dtypes: float64(2), int64(1), object(7)
memory usage: 2.3+ KB


Unnamed: 0  BR  Bat   Ground    Margin      Opposition Result   Team  Toss  \
0         NaN NaN  2nd   Kanpur    5 runs  v South Africa   lost  India  lost   
1         NaN NaN  1st   Indore   22 runs  v South Africa    won  India   won   
2         NaN NaN  2nd   Rajkot   18 runs  v South Africa   lost  India  lost   
3         NaN NaN  1st  Chennai   35 runs  v South Africa    won  India   won   
4         NaN NaN  2nd   Mumbai  214 runs  v South Africa   lost  India  lost   

   Year  
0  2015  
1  2015  
2  2015  
3  2015  
4  2015

In [919]:
Ind_Aus_host_India_ten_years.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18 entries, 0 to 17
Data columns (total 10 columns):
Unnamed: 0    0 non-null float64
BR            9 non-null float64
Bat           18 non-null object
Ground        18 non-null object
Margin        18 non-null object
Opposition    18 non-null object
Result        18 non-null object
Team          18 non-null object
Toss          18 non-null object
Year          18 non-null int64
dtypes: float64(2), int64(1), object(7)
memory usage: 1.5+ KB


In [920]:
Ind_Aus_host_India_selected_venues.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45 entries, 0 to 44
Data columns (total 10 columns):
Unnamed: 0    0 non-null float64
BR            25 non-null float64
Bat           45 non-null object
Ground        45 non-null object
Margin        44 non-null object
Opposition    45 non-null object
Result        45 non-null object
Team          45 non-null object
Toss          45 non-null object
Year          45 non-null int64
dtypes: float64(2), int64(1), object(7)
memory usage: 3.6+ KB


In [921]:
Ind_last_2years.head()

Unnamed: 0    BR  Bat      Ground     Margin      Opposition Result   Team  \
0         NaN   NaN  1st  Birmingham   124 runs      v Pakistan    won  India   
1         NaN   8.0  1st    The Oval  7 wickets     v Sri Lanka   lost  India   
2         NaN  72.0  2nd    The Oval  8 wickets  v South Africa    won  India   
3         NaN  59.0  2nd  Birmingham  9 wickets    v Bangladesh    won  India   
4         NaN   NaN  2nd    The Oval   180 runs      v Pakistan   lost  India   

   Toss  Year  
0  lost  2017  
1  lost  2017  
2   won  2017  
3   won  2017  
4   won  2017

In [922]:
testCases.head()

Bat              Ground Opposition  Toss  Year
0  1st  Hyderabad (Deccan)  Australia   won  2019
1  1st  Hyderabad (Deccan)  Australia  lost  2019
2  2nd  Hyderabad (Deccan)  Australia   won  2019
3  2nd  Hyderabad (Deccan)  Australia  lost  2019
4  1st              Ranchi  Australia   won  2019

In [923]:
Ind_master_df = Ind_last_4years_home.append([Ind_Aus_host_India_ten_years, Ind_Aus_host_India_selected_venues,Ind_last_2years])

In [924]:
Ind_master_df.reset_index()

index  Unnamed: 0     BR  Bat              Ground     Margin  \
0        0         NaN    NaN  2nd              Kanpur     5 runs   
1        1         NaN    NaN  1st              Indore    22 runs   
2        2         NaN    NaN  2nd              Rajkot    18 runs   
3        3         NaN    NaN  1st             Chennai    35 runs   
4        4         NaN    NaN  2nd              Mumbai   214 runs   
5        5         NaN  101.0  2nd          Dharamsala  6 wickets   
6        6         NaN    NaN  2nd               Delhi     6 runs   
7        7         NaN   10.0  2nd              Mohali  7 wickets   
8        8         NaN    NaN  2nd              Ranchi    19 runs   
9        9         NaN    NaN  1st       Visakhapatnam   190 runs   
10      10         NaN   11.0  2nd                Pune  3 wickets   
11      11         NaN    NaN  1st             Cuttack    15 runs   
12      12         NaN    NaN  2nd             Kolkata     5 runs   
13      13         NaN    NaN  1st             Chennai    26 runs   
14      14         NaN    NaN  1st             Kolkata    50 runs   
15      15         NaN   13.0  2nd              Indore  5 wickets   
16      16         NaN    NaN  2nd           Bengaluru    21 runs   
17      17         NaN   43.0  2nd              Nagpur  7 wickets   
18      18         NaN    6.0  1st              Mumbai  6 wickets   
19      19         NaN   24.0  2nd                Pune  6 wickets   
20      20         NaN    NaN  1st              Kanpur     6 runs   
21      21         NaN  176.0  1st          Dharamsala  7 wickets   
22      22         NaN    NaN  1st              Mohali   141 runs   
23      23         NaN  107.0  2nd       Visakhapatnam  8 wickets   
24      24         NaN   47.0  2nd            Guwahati  8 wickets   
25      25         NaN    NaN  1st       Visakhapatnam          -   
26      26         NaN    NaN  2nd                Pune    43 runs   
27      27         NaN    NaN  1st         Mumbai (BS)   224 runs   
28      28         NaN  211.0  2nd  Thiruvananthapuram  9 wickets   
29       0         NaN    NaN  2nd            Vadodara     4 runs   
..     ...         ...    ...  ...                 ...        ...   
112     20         NaN   24.0  2nd                Pune  6 wickets   
113     21         NaN    NaN  1st              Kanpur     6 runs   
114     22         NaN  176.0  1st          Dharamsala  7 wickets   
115     23         NaN    NaN  1st              Mohali   141 runs   
116     24         NaN  107.0  2nd       Visakhapatnam  8 wickets   
117     25         NaN   27.0  2nd              Durban  6 wickets   
118     26         NaN  177.0  2nd           Centurion  9 wickets   
119     27         NaN    NaN  1st           Cape Town   124 runs   
120     28         NaN   15.0  1st        Johannesburg  5 wickets   
121     29         NaN    NaN  1st      Port Elizabeth    73 runs   
122     30         NaN  107.0  2nd           Centurion  8 wickets   
123     31         NaN   59.0  2nd          Nottingham  8 wickets   
124     32         NaN    NaN  2nd              Lord's    86 runs   
125     33         NaN   33.0  1st               Leeds  8 wickets   
126     34         NaN    NaN  1st         Dubai (DSC)    26 runs   
127     35         NaN  126.0  2nd         Dubai (DSC)  8 wickets   
128     36         NaN   82.0  2nd         Dubai (DSC)  7 wickets   
129     37         NaN   63.0  2nd         Dubai (DSC)  9 wickets   
130     38         NaN    NaN  2nd         Dubai (DSC)          -   
131     39         NaN    0.0  2nd         Dubai (DSC)  3 wickets   
132     40         NaN   47.0  2nd            Guwahati  8 wickets   
133     41         NaN    NaN  1st       Visakhapatnam          -   
134     42         NaN    NaN  2nd                Pune    43 runs   
135     43         NaN    NaN  1st         Mumbai (BS)   224 runs   
136     44         NaN  211.0  2nd  Thiruvananthapuram  9 wickets   
137     45         NaN    NaN  2nd              Sydney    34 runs   
138

In [925]:

Ind_master_df.drop_duplicates( inplace=True)

In [926]:
Ind_master_df.reset_index()
print(Ind_master_df.columns)
Ind_master_df.drop(['Unnamed: 0', 'BR', 'Margin'], axis=1, inplace=True)


Index(['Unnamed: 0', 'BR', 'Bat', 'Ground', 'Margin', 'Opposition', 'Result',
       'Team', 'Toss', 'Year'],
      dtype='object')


In [927]:
Ind_master_df.head()

Bat   Ground      Opposition Result   Team  Toss  Year
0  2nd   Kanpur  v South Africa   lost  India  lost  2015
1  1st   Indore  v South Africa    won  India   won  2015
2  2nd   Rajkot  v South Africa   lost  India  lost  2015
3  1st  Chennai  v South Africa    won  India   won  2015
4  2nd   Mumbai  v South Africa   lost  India  lost  2015

In [928]:
Ind_master_df['Opposition'] = Ind_master_df['Opposition'].map(lambda x: x[2:])

In [929]:
Ind_master_df.loc[:,'FormIndicator']  = Ind_master_df['Year'].map(lambda x:'High' if x >= 2018 else ('Medium'if x >= 2015 else 'Low'))

In [930]:
testCases.loc[:,'FormIndicator']  = testCases['Year'].map(lambda x:'High' if x >= 2018 else ('Medium'if x >= 2015 else 'Low'))

In [931]:
Ind_master_df.drop(['Team','Year'],axis=1,inplace=True)
testCases.drop(['Year'],axis=1,inplace=True)
Ind_master_df = Ind_master_df.append([testCases])
Ind_master_df.reset_index()

index  Bat FormIndicator              Ground    Opposition Result  Toss
0        0  2nd        Medium              Kanpur  South Africa   lost  lost
1        1  1st        Medium              Indore  South Africa    won   won
2        2  2nd        Medium              Rajkot  South Africa   lost  lost
3        3  1st        Medium             Chennai  South Africa    won   won
4        4  2nd        Medium              Mumbai  South Africa   lost  lost
5        5  2nd        Medium          Dharamsala   New Zealand    won   won
6        6  2nd        Medium               Delhi   New Zealand   lost   won
7        7  2nd        Medium              Mohali   New Zealand    won   won
8        8  2nd        Medium              Ranchi   New Zealand   lost  lost
9        9  1st        Medium       Visakhapatnam   New Zealand    won   won
10      10  2nd        Medium                Pune       England    won   won
11      11  1st        Medium             Cuttack       England    won  lost
12      12  2nd        Medium             Kolkata       England   lost   won
13      13  1st        Medium             Chennai     Australia    won   won
14      14  1st        Medium             Kolkata     Australia    won   won
15      15  2nd        Medium              Indore     Australia    won  lost
16      16  2nd        Medium           Bengaluru     Australia   lost  lost
17      17  2nd        Medium              Nagpur     Australia    won  lost
18      18  1st        Medium              Mumbai   New Zealand   lost   won
19      19  2nd        Medium                Pune   New Zealand    won  lost
20      20  1st        Medium              Kanpur   New Zealand    won  lost
21      21  1st        Medium          Dharamsala     Sri Lanka   lost  lost
22      22  1st        Medium              Mohali     Sri Lanka    won  lost
23      23  2nd        Medium       Visakhapatnam     Sri Lanka    won   won
24      24  2nd          High            Guwahati   West Indies    won   won
25      25  1st          High       Visakhapatnam   West Indies   tied   won
26      26  2nd          High                Pune   West Indies   lost   won
27      27  1st          High         Mumbai (BS)   West Indies    won   won
28      28  2nd          High  Thiruvananthapuram   West Indies    won  lost
29       0  2nd           Low            Vadodara     Australia   lost  lost
..     ...  ...           ...                 ...           ...    ...   ...
102     35  2nd          High         Dubai (DSC)      Pakistan    won  lost
103     36  2nd          High         Dubai (DSC)    Bangladesh    won   won
104     37  2nd          High         Dubai (DSC)      Pakistan    won  lost
105     38  2nd          High         Dubai (DSC)   Afghanistan   tied  lost
106     39  2nd          High         Dubai (DSC)    Bangladesh    won   won
107     45  2nd          High              Sydney     Australia   lost  lost
108     46  2nd          High            Adelaide     Australia    won  lost
109     47  2nd          High           Melbourne     Australia    won   won
110     48  2nd          High              Napier   New Zealand    won  lost
111     49  1st          High     Mount Maunganui   New Zealand    won   won
112      0  1st          High  Hyderabad (Deccan)     Australia    NaN   won
113      1  1st          High  Hyderabad (Deccan)     Australia    NaN  lost
114      2  2nd          High  Hyderabad (Deccan)     Australia    NaN   won
115      3  2nd          High  Hyderabad (Deccan)     Australia    NaN  lost
116      4  1st          High              Ranchi     Australia    NaN   won
117      5  1st          High              Ranchi     Australia    NaN  lost
118      6  2nd          High              Ranchi     Australia    NaN   won
119      7  2nd          High              Ranchi     Australia    NaN  lost
120      8  1st          High               Delhi     Australia    NaN   won
121      9  1st          High               Delhi     Australia    NaN  lost
1

In [932]:

Ind_master_df_Opposition = pd.get_dummies(Ind_master_df['Opposition'],prefix="Opposition")
#Adding the results to the master dataframe
Ind_master_df = pd.concat([Ind_master_df,Ind_master_df_Opposition],axis=1)
Ind_master_df.drop(['Opposition'],axis=1,inplace=True)

Ind_master_df_Result = pd.get_dummies(Ind_master_df['Result'],prefix="Result")
#Adding the results to the master dataframe
Ind_master_df = pd.concat([Ind_master_df,Ind_master_df_Result],axis=1)
Ind_master_df.drop(['Result'],axis=1,inplace=True)

Ind_master_df_Ground = pd.get_dummies(Ind_master_df['Ground'],prefix="Ground")
#Adding the results to the master dataframe
Ind_master_df = pd.concat([Ind_master_df,Ind_master_df_Ground],axis=1)
Ind_master_df.drop(['Ground'],axis=1,inplace=True)

Ind_master_df_Bat = pd.get_dummies(Ind_master_df['Bat'],prefix="Bat")
#Adding the results to the master dataframe
Ind_master_df = pd.concat([Ind_master_df,Ind_master_df_Bat],axis=1)
Ind_master_df.drop(['Bat'],axis=1,inplace=True)

Ind_master_df_FormIndicator = pd.get_dummies(Ind_master_df['FormIndicator'],prefix="FormIndicator")
#Adding the results to the master dataframe
Ind_master_df = pd.concat([Ind_master_df,Ind_master_df_FormIndicator],axis=1)
Ind_master_df.drop(['FormIndicator'],axis=1,inplace=True)

Ind_master_df_Toss = pd.get_dummies(Ind_master_df['Toss'],prefix="Toss")
#Adding the results to the master dataframe
Ind_master_df = pd.concat([Ind_master_df,Ind_master_df_Toss],axis=1)
Ind_master_df.drop(['Toss'],axis=1,inplace=True)



In [933]:
print(Ind_master_df.head())
testCases.head()

   Opposition_Afghanistan  Opposition_Australia  Opposition_Bangladesh  \
0                       0                     0                      0   
1                       0                     0                      0   
2                       0                     0                      0   
3                       0                     0                      0   
4                       0                     0                      0   

   Opposition_England  Opposition_Hong Kong  Opposition_Netherlands  \
0                   0                     0                       0   
1                   0                     0                       0   
2                   0                     0                       0   
3                   0                     0                       0   
4                   0                     0                       0   

   Opposition_New Zealand  Opposition_Pakistan  Opposition_South Africa  \
0                       0                    0       

Bat              Ground Opposition  Toss FormIndicator
0  1st  Hyderabad (Deccan)  Australia   won          High
1  1st  Hyderabad (Deccan)  Australia  lost          High
2  2nd  Hyderabad (Deccan)  Australia   won          High
3  2nd  Hyderabad (Deccan)  Australia  lost          High
4  1st              Ranchi  Australia   won          High

In [934]:
Ind_master_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 132 entries, 0 to 19
Data columns (total 66 columns):
Opposition_Afghanistan       132 non-null uint8
Opposition_Australia         132 non-null uint8
Opposition_Bangladesh        132 non-null uint8
Opposition_England           132 non-null uint8
Opposition_Hong Kong         132 non-null uint8
Opposition_Netherlands       132 non-null uint8
Opposition_New Zealand       132 non-null uint8
Opposition_Pakistan          132 non-null uint8
Opposition_South Africa      132 non-null uint8
Opposition_Sri Lanka         132 non-null uint8
Opposition_West Indies       132 non-null uint8
Opposition_Zimbabwe          132 non-null uint8
Result_lost                  132 non-null uint8
Result_tied                  132 non-null uint8
Result_won                   132 non-null uint8
Ground_Adelaide              132 non-null uint8
Ground_Ahmedabad             132 non-null uint8
Ground_Bengaluru             132 non-null uint8
Ground_Birmingham            132

In [935]:
# Importing test_train_split from sklearn library
from sklearn.model_selection import train_test_split

In [936]:
# Putting feature variable to X
X_train = Ind_master_df[0:112].drop(['Result_won','Result_lost','Result_tied'],axis=1)

# Putting response variable to y
y_train = Ind_master_df[0:112]['Result_won']

X_test = Ind_master_df[112:].drop(['Result_won','Result_lost','Result_tied'],axis=1)

# Putting response variable to y
y_test = Ind_master_df[112:]['Result_won']

print(y_test)
# Splitting the data into train and test
#X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=101)

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
16    0
17    0
18    0
19    0
Name: Result_won, dtype: uint8


In [937]:
# Importing random forest classifier from sklearn library
from sklearn.ensemble import RandomForestClassifier

# Running the random forest with default parameters.
rfc = RandomForestClassifier()
# fit
rfc.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [938]:
# Making predictions
predictions = rfc.predict(X_test)
India_Win_Percentage = 100*(sum(predictions)/20)
print(India_Win_Percentage)


[1 1 0 0 0 1 1 0 1 1 0 1 1 1 1 1 0 1 0 1]
65.0


#### Run 1 - 70%
#### Run 2 - 80%
#### Run 3 - 90%
#### Run 4 - 75%
#### Run 5 - 95%
#### Run 6 - 80%
#### Run 7 - 60%
#### Run 8 -  70%
#### Run 9 - 75%
#### Run 10 - 65%

#### Average of 10 runs of the algorithm i.e India's percentage of winning = 76%